File: src/dlightrag/adapters/http/rest/routes/model_catalogue.py
# Copyright 2025-2026 Hanlian Lu. SPDX-License-Identifier: Apache-2.0
"""REST adapter for the runtime model catalogue application interface."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Header, HTTPException, Query, Request, Response
from pydantic import BaseModel, ConfigDict

from dlightrag.adapters.http.rest.auth import get_current_user
from dlightrag.adapters.http.rest.routes.deps import enforce_access, get_application
from dlightrag.application.access import AccessAction, UserContext
from dlightrag.application.model_catalogue import (
    ModelCatalogueEntryNotFoundError,
    ModelCatalogueReadOnlyError,
    ModelCatalogueRevisionConflict,
    ModelCatalogueUnavailableError,
    ModelCatalogueValidationError,
    ModelCatalogueView,
)

router = APIRouter()


class ReasoningLevelsPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    off: str | None
    minimal: str | None
    low: str | None
    medium: str | None
    high: str | None
    xhigh: str | None
    max: str | None


class ReasoningProfilePayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    format: str
    levels: ReasoningLevelsPayload


class ModelProfilePayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    context_window_tokens: int
    max_input_tokens: int | None
    max_output_tokens: int | None
    supports_images: bool
    reasoning: ReasoningProfilePayload | None


class ModelCatalogueEntryPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    provider: str
    model: str
    base_url: str | None
    profile: ModelProfilePayload


class ModelCatalogueEntryResponse(ModelCatalogueEntryPayload):
    source: str


class ModelCatalogueResponse(BaseModel):
    revision: str
    models: list[ModelCatalogueEntryResponse]


def _etag(revision: str) -> str:
    return f'"{revision}"'


def _if_match(value: str) -> str:
    normalized = value.strip()
    if normalized.startswith("W/"):
        normalized = normalized[2:].strip()
    if len(normalized) >= 2 and normalized[0] == normalized[-1] == '"':
        normalized = normalized[1:-1]
    return normalized


def _response(view: ModelCatalogueView) -> ModelCatalogueResponse:
    return ModelCatalogueResponse(
        revision=view.revision,
        models=[
            ModelCatalogueEntryResponse(
                provider=item.provider,
                model=item.model,
                base_url=item.base_url,
                profile=ModelProfilePayload.model_validate(item.profile),
                source=item.source,
            )
            for item in view.models
        ],
    )


@router.get("/models/catalogue", response_model=ModelCatalogueResponse)
async def read_model_catalogue(
    request: Request,
    response: Response,
    _user: UserContext = Depends(get_current_user),
) -> ModelCatalogueResponse:
    try:
        view = get_application(request).model_catalogue.read()
    except ModelCatalogueUnavailableError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from None
    response.headers["ETag"] = _etag(view.revision)
    return _response(view)


@router.put("/models/catalogue", response_model=ModelCatalogueResponse)
async def upsert_model_catalogue_entry(
    payload: ModelCatalogueEntryPayload,
    request: Request,
    response: Response,
    if_match: str = Header(alias="If-Match"),
    user: UserContext = Depends(get_current_user),
) -> ModelCatalogueResponse:
    await enforce_access(request, user, AccessAction.MODEL_CATALOGUE_WRITE)
    try:
        view = await get_application(request).model_catalogue.upsert(
            payload.model_dump(),
            expected_revision=_if_match(if_match),
            actor=user.user_id,
        )
    except ModelCatalogueRevisionConflict as exc:
        raise HTTPException(
            status_code=412,
            detail=str(exc),
            headers={"ETag": _etag(exc.current_revision)},
        ) from None
    except ModelCatalogueReadOnlyError as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from None
    except ModelCatalogueUnavailableError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from None
    except ModelCatalogueValidationError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from None
    response.headers["ETag"] = _etag(view.revision)
    return _response(view)


@router.delete("/models/catalogue", response_model=ModelCatalogueResponse)
async def remove_model_catalogue_entry(
    request: Request,
    response: Response,
    provider: str = Query(),
    model: str = Query(),
    base_url: str | None = Query(default=None),
    if_match: str = Header(alias="If-Match"),
    user: UserContext = Depends(get_current_user),
) -> ModelCatalogueResponse:
    await enforce_access(request, user, AccessAction.MODEL_CATALOGUE_WRITE)
    try:
        view = await get_application(request).model_catalogue.remove(
            provider=provider,
            model=model,
            base_url=base_url,
            expected_revision=_if_match(if_match),
            actor=user.user_id,
        )
    except ModelCatalogueRevisionConflict as exc:
        raise HTTPException(
            status_code=412,
            detail=str(exc),
            headers={"ETag": _etag(exc.current_revision)},
        ) from None
    except ModelCatalogueEntryNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc.args[0])) from None
    except ModelCatalogueReadOnlyError as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from None
    except ModelCatalogueUnavailableError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from None
    except ModelCatalogueValidationError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from None
    response.headers["ETag"] = _etag(view.revision)
    return _response(view)


__all__ = [
    "ModelCatalogueEntryPayload",
    "ModelCatalogueResponse",
    "router",
]

File: frontend/api/model_catalogue.ts
// Copyright 2025-2026 Hanlian Lu. SPDX-License-Identifier: Apache-2.0
/** Web API client for runtime model catalogue administration. */

import {csrfHeaders} from './csrf.ts';

export type ReasoningLevel = 'off' | 'minimal' | 'low' | 'medium' | 'high' | 'xhigh' | 'max';

export interface ReasoningProfilePayload {
  format: string;
  levels: Record<ReasoningLevel, string | null>;
}

export interface ModelProfilePayload {
  context_window_tokens: number;
  max_input_tokens: number | null;
  max_output_tokens: number | null;
  supports_images: boolean;
  reasoning: ReasoningProfilePayload | null;
}

export interface ModelCatalogueEntry {
  provider: string;
  model: string;
  base_url: string | null;
  profile: ModelProfilePayload;
  source: 'builtin' | 'overlay' | 'fallback';
}

export interface ModelCatalogue {
  revision: string;
  models: ModelCatalogueEntry[];
}

export class ModelCatalogueRequestError extends Error {
  constructor(readonly status: number, message: string) {
    super(message);
  }
}

async function readResponse(response: Response): Promise<ModelCatalogue> {
  if (!response.ok) {
    let detail = `Model catalogue request failed (${response.status})`;
    try {
      const body = await response.json() as {detail?: string};
      if (body.detail) detail = body.detail;
    } catch {
      // Preserve the bounded status message when the response is not JSON.
    }
    throw new ModelCatalogueRequestError(response.status, detail);
  }
  return await response.json() as ModelCatalogue;
}

export async function getModelCatalogue(signal?: AbortSignal): Promise<ModelCatalogue> {
  return await readResponse(await fetch('/web/api/models/catalogue', {signal}));
}

export async function putModelCatalogueEntry(
  entry: Omit<ModelCatalogueEntry, 'source'>,
  revision: string,
  signal?: AbortSignal,
): Promise<ModelCatalogue> {
  return await readResponse(await fetch('/web/api/models/catalogue', {
    method: 'PUT',
    headers: {...csrfHeaders('application/json'), 'If-Match': revision},
    body: JSON.stringify(entry),
    signal,
  }));
}

export async function removeModelCatalogueEntry(
  entry: Pick<ModelCatalogueEntry, 'provider' | 'model' | 'base_url'>,
  revision: string,
  signal?: AbortSignal,
): Promise<ModelCatalogue> {
  const params = new URLSearchParams({provider: entry.provider, model: entry.model});
  if (entry.base_url) params.set('base_url', entry.base_url);
  return await readResponse(await fetch(`/web/api/models/catalogue?${params}`, {
    method: 'DELETE',
    headers: {...csrfHeaders(), 'If-Match': revision},
    signal,
  }));
}
